bob-apps/bin/bobc.rs

```diff
@@ -1,71 +1,365 @@
+                    let keys = vec.into_iter().map(|v| get_key_value(v, keysize)).collect();
+                    exist(keys, addr).await;
+                }
             }
-            _ => {}
         }
+        _ => unreachable!(),
     }
 }
 
-async fn put(key: Vec<u8>, size: usize, addr: Uri) {
-    let mut client = BobApiClient::connect(addr).await.unwrap();
+async fn prepare_put_from_pattern(re_path: &Regex, dir: &PathBuf, key_size: usize) -> Vec<KeyName> {
+    let mut dir_iter = read_dir(dir).await.unwrap();
 
-    let timestamp = SystemTime::now()
-        .duration_since(UNIX_EPOCH)
-        .expect("msg: &str")
-        .as_secs();
-    let meta = BlobMeta { timestamp };
-    let blob = Blob {
-        data: vec![1; size],
-        meta: Some(meta),
-    };
-    let message = PutRequest {
-        key: Some(BlobKey { key }),
-        data: Some(blob),
-        options: None,
-    };
-    let put_req = Request::new(message);
+    let mut keys_names = Vec::new();
+    while let Some(entry) = dir_iter.next_entry().await.unwrap() {
+        let file = entry.path();
+            let name = file.to_str().unwrap().to_owned();
+            if let (false, Some(cap)) = (file.is_dir(), re_path.captures(&name)) {
+                let key = get_key_value(cap[1].parse().unwrap(), key_size);
+                keys_names.push(KeyName { key, name });
+            }
+    }
+    keys_names
+}
 
-    let res = client.put(put_req).await;
-    info!("{:#?}", res);
+fn prepare_get(filenames: Vec<(u64, String)>, key_size: usize) -> Vec<KeyName> {
+    let mut keys_names = Vec::with_capacity(filenames.len());
+    for (i, name) in filenames {
+        keys_names.push(KeyName {
+            key: get_key_value(i, key_size),
+            name,
+        })
+    }
+    keys_names
 }
 
-async fn get(key: Vec<u8>, addr: Uri) {
+async fn put(key: Vec<u8>, filename: &str, addr: Uri) {
+    let file = File::open(filename).await;
+
+    match file {
+        Err(e) => {
+            error!("{:?}", e);
+        }
+        Ok(mut f) => {
+            let mut client = BobApiClient::connect(addr).await.unwrap();
+
+            let timestamp = SystemTime::now()
+                .duration_since(UNIX_EPOCH)
+                .unwrap()
+                .as_secs();
+            let meta = BlobMeta { timestamp };
+
+            let mut data = Vec::new();
+            match f.read_to_end(&mut data).await {
+                Ok(_) => {}
+                Err(e) => {
+                    error!("{:?}", e);
+                    return;
+                }
+            }
+            let blob = Blob {
+                data,
+                meta: Some(meta),
+            };
+            let message = PutRequest {
+                key: Some(BlobKey { key }),
+                data: Some(blob),
+                options: None,
+            };
+            let put_req = Request::new(message);
+
+            let res = client.put(put_req).await;
+            info!("{:#?}", res);
```

Comment:
You write log for every record 2 times: first in the caller and second - here. It is better to write only one line for record in the caller and here just return some result to incorporate it into a message on the caller side

Reply:
Done: [eda6196](https://github.com/qoollo/bob/pull/562/commits/eda6196444d62621a945784518d4230088266869)